src/tasks/publish/index.ts

```diff
@@ -0,0 +1,226 @@
+    managerAddress,
+    ipfsProvider,
+    onlyContent,
+    noVerify
+  }: {
+    bumpOrVersion: string
+    existingContractAddress: string
+    managerAddress: string
+    ipfsProvider: string
+    onlyContent: boolean
+    noVerify: boolean
+  },
+  bre: BuidlerRuntimeEnvironment
+): Promise<void> {
+  const aragonConfig = bre.config.aragon as AragonConfig
+  const appSrcPath = aragonConfig.appSrcPath as string
+  const distPath = aragonConfig.appBuildOutputPath as string
+  const ipfsIgnore = ['subfolder/to/ignore/**'] // Hardcoded until a better way to deal with dynamic ENS address is found
+  const selectedNetwork = bre.network.name
+  // ### Todo: Add logic to guess app name from file?
+  const appName =
+    typeof aragonConfig.appName === 'string'
+      ? aragonConfig.appName
+      : typeof aragonConfig.appName === 'object'
+      ? aragonConfig.appName[selectedNetwork]
+      : undefined
+  if (!appName)
+    throw new BuidlerPluginError(`appName must be defined in buidler.config`)
+  const contractName = getMainContractName()
+
+  // Initialize clients
+  const networkConfig = bre.network.config as HttpNetworkConfig
+  const provider = new ethers.providers.Web3Provider(
+    bre.web3.currentProvider,
+    networkConfig.ensAddress && {
+      name: bre.network.name,
+      chainId: networkConfig.chainId || 5555,
+      ensAddress: networkConfig.ensAddress
+    }
+  )
+
+  const prevVersion = await _getLastestVersionIfExists(appName, provider)
+
+  const { bump, nextVersion } = parseAndValidateBumpOrVersion(
+    bumpOrVersion,
+    prevVersion ? prevVersion.version : undefined
+  )
+  logMain(`Applying version bump ${bump}, next version: ${nextVersion}`)
+
+  // Using let + if {} block instead of a ternary operator
+  // to assign value and log status to console
+  let contractAddress: string
+  if (onlyContent) {
+    logMain('No contract used for this version')
+    contractAddress = zeroAddress
+  } else if (existingContractAddress) {
+    logMain('Using provided contract address')
+    contractAddress = existingContractAddress
+  } else if (!prevVersion || bump === 'major') {
+    logMain('Deploying new implementation contract')
+    contractAddress = await _deployMainContract(contractName, noVerify, bre)
+  } else {
+    logMain('Reusing contract from previous version')
+    contractAddress = prevVersion.contractAddress
+  }
+  logMain(`contractAddress: ${contractAddress}`)
+
+  // Prepare release directory
+  // npm run build must create: index.html, src.js, script.js
+  logMain(`Building app front-end at ${appSrcPath}`)
+  await execa('npm', ['run', 'build'], { cwd: appSrcPath })
+
+  // Generate and validate Aragon artifacts, release files
+  logMain(`Generating Aragon app artifacts`)
+  await generateArtifacts(distPath, bre)
+  validateArtifacts(distPath)
+
+  // Upload release directory to IPFS
+  logMain('Uploading release assets to IPFS...')
+  const contentHash = await uploadReleaseToIpfs(distPath, {
```

Comment:
This assumes there's an IPFS daemon running at the specified location, right? We should probably sanity check that one exists before getting here.

Reply:
No, it does a POST to the provided IPFS API URL with the files

Reply:
But before we do that POST, we should probably double check that IPFS API is responsive

Reply:
Added on #77